polar-core/src/partial/simplify.rs

```diff
@@ -94,42 +94,117 @@ pub fn simplify_partial(var: &Symbol, term: Term) -> Term {
 ///
 /// - For partials, simplify the constraint expressions.
 /// - For non-partials, deep deref. TODO(ap/gj): deep deref.
-pub fn simplify_bindings(bindings: Bindings) -> Option<Bindings> {
+pub fn simplify_bindings(bindings: Bindings, all: bool) -> Option<Bindings> {
     let mut unsatisfiable = false;
     let mut simplify = |var: Symbol, term: Term| {
         let simplified = simplify_partial(&var, term);
         match simplified.value().as_expression() {
             Ok(o) if o == &FALSE => unsatisfiable = true,
             _ => (),
         }
-        simplified
+        let mut symbols = HashSet::new();
+        simplified.variables(&mut symbols);
+        (simplified, symbols)
     };
 
-    let bindings: Bindings = bindings
-        .iter()
-        .map(|(var, value)| match value.value() {
-            Value::Expression(o) => {
-                assert_eq!(o.operator, Operator::And);
-                (var.clone(), simplify(var.clone(), value.clone()))
+    let mut simplified_bindings = HashMap::new();
+    if all {
+        for (var, value) in &bindings {
+            match value.value() {
+                Value::Expression(o) => {
+                    assert_eq!(o.operator, Operator::And);
+                    let (simplified, _) = simplify(var.clone(), value.clone());
+                    simplified_bindings.insert(var.clone(), simplified)
+                }
+                Value::Variable(v) | Value::RestVariable(v)
+                    if v.is_temporary_var()
+                        && bindings.contains_key(v)
+                        && matches!(
+                            bindings[v].value(),
+                            Value::Variable(_) | Value::RestVariable(_)
+                        ) =>
+                {
+                    simplified_bindings.insert(var.clone(), bindings[v].clone())
+                }
+                _ => simplified_bindings.insert(var.clone(), value.clone()),
+            };
+        }
+    } else {
+        let mut referenced_vars: VecDeque<Symbol> = VecDeque::new();
+        for (var, value) in &bindings {
+            if !var.is_temporary_var() {
+                match value.value() {
```

Comment:
This & the below could probably be a `simplify_var` function that just returns the simplified var & the referenced variables, right? Or returns the simplified var and getting the referenced vars is a separate step from it that only happens in the `!all` case.